### PR TITLE
[codex] refine shell prompt and powershell invocation

### PR DESCRIPTION
**Pull request by aboimpinto**
## Summary

- refine the Windows shell prompt so it shows a shell stem instead of a full path
- add `-NonInteractive` to PowerShell invocation
- keep the `cmd.exe` full-path raw-argument handling fix and its regression coverage
- keep the `shell_program_stem` helper reusable inside the crate

This is a follow-up branch because I cannot push directly to the maintainer branch from this environment.

Paulo Aboim Pinto

<details open><summary><h3>Greptile Summary</h3></summary>

This PR refines Windows shell handling in three focused ways: the system-prompt environment block now shows a shell stem (`pwsh`, `cmd`) instead of a full executable path; PowerShell invocations gain `-NonInteractive`; and `push_eval_shell_args` in `eval.rs` correctly matches full-path `cmd.exe` references by routing through `shell_program_stem` rather than a plain case-insensitive string compare.

- The removal of the implicit `powershell.exe` (Windows PowerShell 5.x) auto-detection fallback is intentional and is now documented with a code comment; users can still select it explicitly via the `SHELL` environment variable.
- `-NonInteractive` is added to every `powershell_invocation` call (both `pwsh.exe` and explicit `powershell.exe` via `SHELL`), preventing PowerShell from waiting for user input in automated contexts.
- `shell_program_stem` is promoted to `pub(crate)` and reused consistently across `eval.rs` and `prompts.rs`, eliminating duplicated path-normalisation logic.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge; all behaviour changes are intentional, documented, and covered by new tests.

The three changed files make tightly scoped improvements. The powershell.exe auto-detection removal is explicitly documented in a code comment and a new test confirms the correct fallback path. The shell_program_stem refactor produces identical output for all edge cases. The -NonInteractive addition and stem-only prompt display have targeted regression tests. No logic regressions or data-correctness issues were identified.

No files require special attention.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| crates/tui/src/shell_invocation.rs | Removes powershell_on_path auto-detection (now documented as intentional); adds -NonInteractive to PowerShell invocation; makes shell_program_stem pub(crate); refactors stem extraction to lowercase upfront; adds three new targeted tests. |
| crates/tui/src/prompts.rs | Uses shell_program_stem on the resolved program to emit only the shell stem instead of a full path in the Environment block; adds a Windows-only assertion to the existing environment-block test. |
| crates/tui/src/eval.rs | Imports shell_program_stem and replaces the direct case-insensitive program name check so full-path cmd.exe invocations correctly use raw args; adds a Windows-only regression test. |

</details>

</details>

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `crates/tui/src/shell_invocation.rs`, line 129-131 ([link](https://github.com/hmbown/codewhale/blob/8e1a27e439aa30b3cc05f005054a89094d15a93c/crates/tui/src/shell_invocation.rs#L129-L131)) 

   <a href="#">P1</a> **Silent removal of `powershell.exe` auto-detection fallback**

   The `powershell_on_path` field and its associated fallback branch were removed, which means Windows users who only have `powershell.exe` (Windows PowerShell 5.x) installed — but not `pwsh` (PowerShell 7) — will now silently fall through to `cmd.exe`. This is a non-trivial behaviour change for any Windows machine that ships PowerShell 5.x exclusively (e.g. Windows Server 2019, unmanaged enterprise desktops). The PR description does not mention this removal, so it is unclear whether it is intentional. If `powershell.exe` is being deliberately dropped, a short comment or changelog entry would help reviewers and users understand the new minimum requirement.

   <a href="https://chatgpt.com/codex/deeplink?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22hmbown%2Fcodewhale%22%20on%20the%20existing%20branch%20%22codex%2F2279-followup%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22codex%2F2279-followup%22.%0A%0AThis%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20crates%2Ftui%2Fsrc%2Fshell_invocation.rs%0ALine%3A%20129-131%0A%0AComment%3A%0A**Silent%20removal%20of%20%60powershell.exe%60%20auto-detection%20fallback**%0A%0AThe%20%60powershell_on_path%60%20field%20and%20its%20associated%20fallback%20branch%20were%20removed%2C%20which%20means%20Windows%20users%20who%20only%20have%20%60powershell.exe%60%20%28Windows%20PowerShell%205.x%29%20installed%20%E2%80%94%20but%20not%20%60pwsh%60%20%28PowerShell%207%29%20%E2%80%94%20will%20now%20silently%20fall%20through%20to%20%60cmd.exe%60.%20This%20is%20a%20non-trivial%20behaviour%20change%20for%20any%20Windows%20machine%20that%20ships%20PowerShell%205.x%20exclusively%20%28e.g.%20Windows%20Server%202019%2C%20unmanaged%20enterprise%20desktops%29.%20The%20PR%20description%20does%20not%20mention%20this%20removal%2C%20so%20it%20is%20unclear%20whether%20it%20is%20intentional.%20If%20%60powershell.exe%60%20is%20being%20deliberately%20dropped%2C%20a%20short%20comment%20or%20changelog%20entry%20would%20help%20reviewers%20and%20users%20understand%20the%20new%20minimum%20requirement.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise."><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodex.svg?v=3">Fix in Codex</picture></a> <a href="https://app.greptile.com/ide/claude-code?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20crates%2Ftui%2Fsrc%2Fshell_invocation.rs%0ALine%3A%20129-131%0A%0AComment%3A%0A**Silent%20removal%20of%20%60powershell.exe%60%20auto-detection%20fallback**%0A%0AThe%20%60powershell_on_path%60%20field%20and%20its%20associated%20fallback%20branch%20were%20removed%2C%20which%20means%20Windows%20users%20who%20only%20have%20%60powershell.exe%60%20%28Windows%20PowerShell%205.x%29%20installed%20%E2%80%94%20but%20not%20%60pwsh%60%20%28PowerShell%207%29%20%E2%80%94%20will%20now%20silently%20fall%20through%20to%20%60cmd.exe%60.%20This%20is%20a%20non-trivial%20behaviour%20change%20for%20any%20Windows%20machine%20that%20ships%20PowerShell%205.x%20exclusively%20%28e.g.%20Windows%20Server%202019%2C%20unmanaged%20enterprise%20desktops%29.%20The%20PR%20description%20does%20not%20mention%20this%20removal%2C%20so%20it%20is%20unclear%20whether%20it%20is%20intentional.%20If%20%60powershell.exe%60%20is%20being%20deliberately%20dropped%2C%20a%20short%20comment%20or%20changelog%20entry%20would%20help%20reviewers%20and%20users%20understand%20the%20new%20minimum%20requirement.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=hmbown%2Fcodewhale&pr=2284&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaude.svg?v=3">Fix in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/cursor?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20crates%2Ftui%2Fsrc%2Fshell_invocation.rs%0ALine%3A%20129-131%0A%0AComment%3A%0A**Silent%20removal%20of%20%60powershell.exe%60%20auto-detection%20fallback**%0A%0AThe%20%60powershell_on_path%60%20field%20and%20its%20associated%20fallback%20branch%20were%20removed%2C%20which%20means%20Windows%20users%20who%20only%20have%20%60powershell.exe%60%20%28Windows%20PowerShell%205.x%29%20installed%20%E2%80%94%20but%20not%20%60pwsh%60%20%28PowerShell%207%29%20%E2%80%94%20will%20now%20silently%20fall%20through%20to%20%60cmd.exe%60.%20This%20is%20a%20non-trivial%20behaviour%20change%20for%20any%20Windows%20machine%20that%20ships%20PowerShell%205.x%20exclusively%20%28e.g.%20Windows%20Server%202019%2C%20unmanaged%20enterprise%20desktops%29.%20The%20PR%20description%20does%20not%20mention%20this%20removal%2C%20so%20it%20is%20unclear%20whether%20it%20is%20intentional.%20If%20%60powershell.exe%60%20is%20being%20deliberately%20dropped%2C%20a%20short%20comment%20or%20changelog%20entry%20would%20help%20reviewers%20and%20users%20understand%20the%20new%20minimum%20requirement.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&pr=2284&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursorDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursor.svg?v=3">Fix in Cursor</picture></a>
</details>

<sub>Reviews (3): Last reviewed commit: ["test(tui): lock Windows shell fallback b..."](https://github.com/hmbown/codewhale/commit/0360192ad72e88d16e2021cca63c74ab1a45d9e0) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=34156950)</sub>

### CODE REVIEW

**Comment by aboimpinto** [comment]
@Hmbown I want to confirm the intended direction before I make another change here.

In your #1779 / #2279 comments, you said that if `pwsh.exe` is not available on `PATH`, it is safer to fall back to `cmd.exe`. The current review on #2284 is asking to restore `powershell_on_path` and bring back a `powershell.exe` fallback.

Can you confirm which behavior you want for this branch?
- `pwsh.exe` -> `powershell.exe` -> `cmd.exe`
- or `pwsh.exe` -> `cmd.exe`

I’ll align the branch to your preference.

Paulo Aboim Pinto

**Comment by Hmbown** [comment]
**pwsh.exe → cmd.exe - thank you so much!**

**Comment by aboimpinto** [comment]
@Hmbown aligned this with your preference: the default Windows fallback stays `pwsh.exe -> cmd.exe`; `powershell.exe` is only used when explicitly selected through `SHELL`.

I also restored coverage for the COMSPEC full-path `cmd.exe` branch in `shell_invocation.rs`, including the UTF-8 prefix and display-command behavior.

Verified locally:
- `cargo fmt --check`
- `cargo test -p codewhale-tui --bin codewhale-tui shell_invocation -- --nocapture`
- `cargo test -p codewhale-tui --bin codewhale-tui push_eval_shell_args -- --nocapture`

Paulo Aboim Pinto

File: crates/tui/src/shell_invocation.rs [hunk 1]
```diff
@@ -75,7 +75,6 @@ pub(crate) struct ShellProbe {
     pub(crate) shell: Option<String>,
     pub(crate) comspec: Option<String>,
     pub(crate) pwsh_on_path: bool,
-    pub(crate) powershell_on_path: bool,
 }
```
[the remark this answers is not in the document]
Comment (aboimpinto) [reply crates/tui/src/shell_invocation.rs:78]: Keeping `powershell_on_path` removed here based on Hmbown's clarification: the intended default Windows resolution is `pwsh.exe -> cmd.exe`.

Windows PowerShell 5.x is still supported when explicitly selected through `SHELL`, but it is not probed as an implicit fallback. Added regression coverage in `0360192a` for both sides of that contract:
- `windows_without_pwsh_falls_straight_to_cmd_not_windows_powershell`
- `windows_shell_env_can_select_windows_powershell`

Paulo Aboim Pinto

File: crates/tui/src/shell_invocation.rs [hunk 2]
```diff
@@ -89,7 +88,6 @@ impl ShellProbe {
                 .ok()
                 .filter(|value| !value.trim().is_empty()),
             pwsh_on_path: command_on_path("pwsh.exe") || command_on_path("pwsh"),
-            powershell_on_path: command_on_path("powershell.exe") || command_on_path("powershell"),
         }
```
[the remark this answers is not in the document]
Comment (aboimpinto) [reply crates/tui/src/shell_invocation.rs:91]: Not restoring automatic `powershell.exe` detection for the same reason: Hmbown confirmed the fallback should be `pwsh.exe -> cmd.exe`.

Explicit `SHELL=powershell.exe` still works, but default probing now intentionally avoids Windows PowerShell 5.x. Commit `0360192a` adds regression tests for both explicit Windows PowerShell selection and the no-`pwsh` direct-to-`cmd` fallback.

Paulo Aboim Pinto

File: crates/tui/src/shell_invocation.rs [hunk 3]
```diff
@@ -132,10 +130,6 @@ fn windows_shell_invocation(command: &str, probe: &ShellProbe) -> ShellInvocatio
         return powershell_invocation("pwsh.exe", command);
     }
 
-    if probe.powershell_on_path {
-        return powershell_invocation("powershell.exe", command);
-    }
-
     if let Some(comspec) = probe
```
[the remark this answers is not in the document]
Comment (aboimpinto) [reply crates/tui/src/shell_invocation.rs:136]: Leaving the `powershell.exe` fallback out intentionally after maintainer confirmation: default Windows fallback is `pwsh.exe -> cmd.exe`.

I added an inline comment in `windows_shell_invocation` to make that decision explicit, kept `powershell.exe` available only through explicit `SHELL` selection, and added regression coverage in `0360192a` to lock both behaviors.

Paulo Aboim Pinto

File: crates/tui/src/shell_invocation.rs [hunk 4]
```diff
@@ -290,31 +283,11 @@ mod tests {
...
-            ["/C", r#"chcp 65001 >NUL & git commit -m "hello world""#]
-        );
-        assert!(invocation.raw_payload_on_windows);
-        assert_eq!(
-            invocation.display_command().as_deref(),
-            Some(r#"git commit -m "hello world""#)
+            ["-NoProfile", "-NonInteractive", "-Command", "Get-ChildItem"]
         );
+        assert!(!invocation.raw_payload_on_windows);
     }
```
[the remark this answers is not in the document]
Comment (aboimpinto) [reply crates/tui/src/shell_invocation.rs:373]: Fixed in `3b66f27b`: restored `windows_falls_back_to_comspec_cmd_with_utf8_prefix` in `shell_invocation.rs`.

The test now covers the full COMSPEC `cmd.exe` path, the UTF-8 prefix, `raw_payload_on_windows`, and `display_command()` stripping. Commit `0360192a` adds the remaining fallback regression coverage for explicit Windows PowerShell selection vs. default no-`pwsh` direct-to-`cmd` behavior.

Paulo Aboim Pinto

[the remark this answers is not in the document]
Comment (aboimpinto) [reply crates/tui/src/shell_invocation.rs:373]: Fixed in `3b66f27b`: `shell_invocation.rs` now directly exercises the COMSPEC full-path `cmd.exe` branch again.

This covers the exact gap you called out: preserving the full `cmd.exe` path, setting `raw_payload_on_windows`, and stripping the UTF-8 prefix from `display_command()`. Commit `0360192a` adds two more regression tests around explicit Windows PowerShell selection and default fallback to `cmd` when `pwsh` is absent.

Paulo Aboim Pinto

**Comment by Hmbown** [comment]
Thank you so much!

**Comment by Hmbown** [comment]
Thanks for continuing the Windows shell work. I am keeping this one open for now because it is still a draft, it targets the #2279 feature branch rather than `main`, and it changes the implicit PowerShell fallback contract more broadly than #1779 needs.

The cleanest path is to settle/rebase #2279 first, then decide whether removing the fallback is an intentional follow-up behavior change. If we do keep this branch, I would like the PR body to spell out the PowerShell 5.x/default-shell behavior and include a Windows check around the raw `cmd.exe` path.

**Comment by Hmbown** [comment]
Thanks @aboimpinto - closing this follow-up as superseded by the ShellDispatcher stack now on main. I do not want to merge it as-is because it reintroduces a second shell abstraction and intentionally drops implicit Windows PowerShell fallback, while current main keeps `SHELL` / `pwsh.exe` / `powershell.exe` / `cmd.exe` resolution in one place.

The landed tests around `shell_dispatcher` and `CommandSpec::shell` are passing locally, so the core Windows quoting issue is covered.
